### ISSUE TITLE
PHP 5.6 - Session lock issue

### ISSUE THREAD

**Issue by ihor-sviziev** [colinmollenhour/Cm_RedisSession#51]
After upgrade to php 5.6 I started see issues with website performance, home page few times loads for 1-1.5 seconds, each next - 35 seconds. 

My investigation result - biggest time exection - redis session, see profiler screenshot below:
![image](https://cloud.githubusercontent.com/assets/1873745/5490373/b4691270-86d9-11e4-9d2e-ab6696779ec7.png)

After enabling debug log level - in redis_session.log I see next messages:
```
2014-12-18T15:13:59+00:00 DEBUG (7): igor|24760: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00058 seconds
2014-12-18T15:13:59+00:00 DEBUG (7): igor|24760: Connected to Redis
2014-12-18T15:13:59+00:00 DEBUG (7): igor|24760: Attempting read lock on ID sess_2t063m3dvskjnqni8cja9biob3
2014-12-18T15:13:59+00:00 DEBUG (7): igor|24760: Data read for ID sess_2t063m3dvskjnqni8cja9biob3 after 0.00077 seconds
2014-12-18T15:14:01+00:00 DEBUG (7): igor|24760: Attempting write to ID 2t063m3dvskjnqni8cja9biob3
2014-12-18T15:14:01+00:00 DEBUG (7): igor|24760: Write lock obtained on ID 2t063m3dvskjnqni8cja9biob3
2014-12-18T15:14:01+00:00 DEBUG (7): igor|24760: Compressing 2079 bytes with lzf
2014-12-18T15:14:01+00:00 DEBUG (7): igor|24760: Data compressed by 73.0 percent in 0.00011 seconds
2014-12-18T15:14:01+00:00 DEBUG (7): igor|24760: Data written to ID 2t063m3dvskjnqni8cja9biob3 after 0.00046 seconds
2014-12-18T15:14:01+00:00 DEBUG (7): igor|24760: Closing connection
2014-12-18T15:14:04+00:00 DEBUG (7): igor|24766: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00049 seconds
2014-12-18T15:14:04+00:00 DEBUG (7): igor|24766: Connected to Redis
2014-12-18T15:14:04+00:00 DEBUG (7): igor|24766: Attempting read lock on ID sess_2t063m3dvskjnqni8cja9biob3
2014-12-18T15:14:04+00:00 DEBUG (7): igor|24766: Data read for ID sess_2t063m3dvskjnqni8cja9biob3 after 0.00044 seconds
2014-12-18T15:14:04+00:00 DEBUG (7): igor|24766: Closing connection
2014-12-18T15:14:06+00:00 DEBUG (7): igor|7572: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00048 seconds
2014-12-18T15:14:06+00:00 DEBUG (7): igor|7572: Connected to Redis
2014-12-18T15:14:06+00:00 DEBUG (7): igor|7572: Attempting read lock on ID sess_2t063m3dvskjnqni8cja9biob3
2014-12-18T15:14:06+00:00 DEBUG (7): igor|7572: Waiting for lock on ID sess_2t063m3dvskjnqni8cja9biob3 (0 tries, 1 waiting, 0.00029 seconds elapsed)
2014-12-18T15:14:17+00:00 DEBUG (7): igor|7572: Checking for zombies after 10.50399 seconds of waiting...
2014-12-18T15:14:27+00:00 DEBUG (7): igor|7572: Checking for zombies after 20.50828 seconds of waiting...
2014-12-18T15:14:37+00:00 DEBUG (7): igor|7572: Checking for zombies after 30.51300 seconds of waiting...
2014-12-18T15:14:41+00:00 INFO (6): igor|7572: Successfully broke lock for ID sess_2t063m3dvskjnqni8cja9biob3 after 34.51524 seconds (33 attempts). Lock: 35, BREAK_MODULO: 5
Last request of broken lock: GET www.domain.local/
2014-12-18T15:14:41+00:00 DEBUG (7): igor|7572: Data read for ID sess_2t063m3dvskjnqni8cja9biob3 after 34.51647 seconds
2014-12-18T15:14:41+00:00 DEBUG (7): igor|7572: Closing connection
2014-12-18T15:20:28+00:00 DEBUG (7): igor|24767: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00048 seconds
2014-12-18T15:20:28+00:00 DEBUG (7): igor|24767: Connected to Redis
2014-12-18T15:20:28+00:00 DEBUG (7): igor|24767: Attempting read lock on ID sess_2t063m3dvskjnqni8cja9biob3
2014-12-18T15:20:28+00:00 DEBUG (7): igor|24767: Waiting for lock on ID sess_2t063m3dvskjnqni8cja9biob3 (0 tries, 1 waiting, 0.00032 seconds elapsed)
2014-12-18T15:20:38+00:00 DEBUG (7): igor|24767: Checking for zombies after 10.50416 seconds of waiting...
2014-12-18T15:20:48+00:00 DEBUG (7): igor|24767: Checking for zombies after 20.50856 seconds of waiting...
2014-12-18T15:20:58+00:00 DEBUG (7): igor|24767: Checking for zombies after 30.51281 seconds of waiting...
2014-12-18T15:21:02+00:00 INFO (6): igor|24767: Successfully broke lock for ID sess_2t063m3dvskjnqni8cja9biob3 after 34.51496 seconds (33 attempts). Lock: 35, BREAK_MODULO: 5
Last request of broken lock: GET www.domain.local/
2014-12-18T15:21:02+00:00 DEBUG (7): igor|24767: Data read for ID sess_2t063m3dvskjnqni8cja9biob3 after 34.51652 seconds
2014-12-18T15:21:03+00:00 DEBUG (7): igor|24767: Closing connection
```

OS: Fedora 21
PHP: 5.6.3
php redis version: 2.2.5

Magento version: 1.9.1

I think it's may be related with this: http://php.net/manual/en/book.session.php#116217

Same project with php 5.5 on Fedora 20 has next log (lock issue isn't present):
```
2014-12-18T15:31:12+00:00 DEBUG (7): somepc|25512: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00047 seconds
2014-12-18T15:31:12+00:00 DEBUG (7): somepc|25512: Connected to Redis
2014-12-18T15:31:12+00:00 DEBUG (7): somepc|25512: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:12+00:00 DEBUG (7): somepc|25512: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 0.00046 seconds
2014-12-18T15:31:16+00:00 DEBUG (7): somepc|26998: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00042 seconds
2014-12-18T15:31:16+00:00 DEBUG (7): somepc|26998: Connected to Redis
2014-12-18T15:31:16+00:00 DEBUG (7): somepc|26998: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:16+00:00 DEBUG (7): somepc|26998: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 0.00035 seconds
2014-12-18T15:31:19+00:00 DEBUG (7): somepc|26512: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00042 seconds
2014-12-18T15:31:19+00:00 DEBUG (7): somepc|26512: Connected to Redis
2014-12-18T15:31:19+00:00 DEBUG (7): somepc|26512: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:19+00:00 DEBUG (7): somepc|26512: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 0.00047 seconds
2014-12-18T15:31:21+00:00 DEBUG (7): somepc|25513: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00042 seconds
2014-12-18T15:31:21+00:00 DEBUG (7): somepc|25513: Connected to Redis
2014-12-18T15:31:21+00:00 DEBUG (7): somepc|25513: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:21+00:00 DEBUG (7): somepc|25513: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 0.00042 seconds
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26668: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00044 seconds
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26668: Connected to Redis
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26668: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26668: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 0.00034 seconds
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|25512: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00045 seconds
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|25512: Connected to Redis
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|25512: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|25512: Waiting for lock on ID sess_pthhijtur71ibaemsaeut0ehl2 (0 tries, 1 waiting, 0.00023 seconds elapsed)
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26511: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00042 seconds
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26511: Connected to Redis
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26511: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:23+00:00 DEBUG (7): somepc|26511: Waiting for lock on ID sess_pthhijtur71ibaemsaeut0ehl2 (0 tries, 2 waiting, 0.00021 seconds elapsed)
2014-12-18T15:31:25+00:00 DEBUG (7): somepc|25512: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 2.50133 seconds
2014-12-18T15:31:26+00:00 DEBUG (7): somepc|26511: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 3.50141 seconds
2014-12-18T15:31:29+00:00 DEBUG (7): somepc|25513: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00042 seconds
2014-12-18T15:31:29+00:00 DEBUG (7): somepc|25513: Connected to Redis
2014-12-18T15:31:29+00:00 DEBUG (7): somepc|25513: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:29+00:00 DEBUG (7): somepc|25513: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 0.00027 seconds
2014-12-18T15:31:30+00:00 DEBUG (7): somepc|25478: Cm_RedisSession_Model_Session initialized for connection to localhost:6379 after 0.00042 seconds
2014-12-18T15:31:30+00:00 DEBUG (7): somepc|25478: Connected to Redis
2014-12-18T15:31:30+00:00 DEBUG (7): somepc|25478: Attempting read lock on ID sess_pthhijtur71ibaemsaeut0ehl2
2014-12-18T15:31:30+00:00 DEBUG (7): somepc|25478: Data read for ID sess_pthhijtur71ibaemsaeut0ehl2 after 0.00034 seconds
```

**Comment by ihor-sviziev** [comment]
@colinmollenhour any updates?

**Comment by colinmollenhour** [comment]
Please try d86c519 (branch php56).

**Comment by ihor-sviziev** [comment]
@colinmollenhour  I'll check it and tell you result.
Thanks!

**Comment by ihor-sviziev** [comment]
Strange... I can't reproduce this issue, now I have php version 5.6.5.

Looks like this issue reproduces because of bug that was fixed in 5.6.4 - https://bugs.php.net/bug.php?id=17860

Can anyone test fixed version on version 5.6.3?
